<a href="https://colab.research.google.com/github/gokulh12/Portfolio/blob/main/MediaMixModeling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import plotly.express as px
import statsmodels.api as sm
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

In [2]:
data = pd.read_csv('dataset_real.csv')
data

ordine_data  revenue  google_search_spent  google_performance_max_spent  \
0    2020-07-01  1878.71                93.04                          0.00   
1    2020-07-02  2057.76                81.20                          0.00   
2    2020-07-03  1930.55                69.40                          0.00   
3    2020-07-04   778.36                34.94                          0.00   
4    2020-07-05  1753.69                90.33                          0.00   
..          ...      ...                  ...                           ...   
591  2022-02-17  6061.37               347.75                        125.32   
592  2022-02-18  3722.15               258.10                         79.30   
593  2022-02-19  2670.47                92.04                          9.96   
594  2022-02-20  3385.87                86.37                         40.50   
595  2022-02-21   372.27                42.17                         22.04   

     fb_retargeting_spent  fb_prospecting_spent  google_organico  Sconti  
0                    0.00                  0.00                0       1  
1                    0.00                  0.00                0       1  
2                    0.00                  0.00                0       2  
3                    0.00                  0.00                0       0  
4                    0.00                  0.00                0       1  
..                    ...                   ...              ...     ...  
591                 19.96                 30.44             5003       9  
592                 20.31                 31.06             4696      14  
593                 23.16                 33.75             3058       8  
594                 19.90                 30.59             1398       3  
595                  1.33                  2.13                0       0  

[596 rows x 8 columns]

In [3]:
data.dtypes

ordine_data                      object
revenue                         float64
google_search_spent             float64
google_performance_max_spent    float64
fb_retargeting_spent            float64
fb_prospecting_spent            float64
google_organico                   int64
Sconti                            int64
dtype: object

In [4]:
data['ordine_data'] = pd.to_datetime(data['ordine_data'])


In [5]:
#transforming date
data['year'] = data['ordine_data'].dt.year
data['month'] = data['ordine_data'].dt.month
data['day_of_week'] = data['ordine_data'].dt.dayofweek
data['day'] = data['ordine_data'].dt.day

In [6]:
data

ordine_data  revenue  google_search_spent  google_performance_max_spent  \
0    2020-07-01  1878.71                93.04                          0.00   
1    2020-07-02  2057.76                81.20                          0.00   
2    2020-07-03  1930.55                69.40                          0.00   
3    2020-07-04   778.36                34.94                          0.00   
4    2020-07-05  1753.69                90.33                          0.00   
..          ...      ...                  ...                           ...   
591  2022-02-17  6061.37               347.75                        125.32   
592  2022-02-18  3722.15               258.10                         79.30   
593  2022-02-19  2670.47                92.04                          9.96   
594  2022-02-20  3385.87                86.37                         40.50   
595  2022-02-21   372.27                42.17                         22.04   

     fb_retargeting_spent  fb_prospecting_spent  google_organico  Sconti  \
0                    0.00                  0.00                0       1   
1                    0.00                  0.00                0       1   
2                    0.00                  0.00                0       2   
3                    0.00                  0.00                0       0   
4                    0.00                  0.00                0       1   
..                    ...                   ...              ...     ...   
591                 19.96                 30.44             5003       9   
592                 20.31                 31.06             4696      14   
593                 23.16                 33.75             3058       8   
594                 19.90                 30.59             1398       3   
595                  1.33                  2.13                0       0   

     year  month  day_of_week  day  
0    2020      7            2    1  
1    2020      7            3    2  
2    2020      7            4    3  
3    2020      7            5    4  
4    2020      7            6    5  
..    ...    ...          ...  ...  
591  2022      2            3   17  
592  2022      2            4   18  
593  2022      2            5   19  
594  2022      2            6   20  
595  2022      2            0   21  

[596 rows x 12 columns]

In [7]:
#checking correlatoin
data.corr()
#We can see some postive correlation for Sconti which refers to discounts

<ipython-input-7-ab0cab1bfc1d>:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.corr()


revenue  google_search_spent  \
revenue                       1.000000             0.158135   
google_search_spent           0.158135             1.000000   
google_performance_max_spent -0.217984             0.253990   
fb_retargeting_spent         -0.212061             0.336720   
fb_prospecting_spent         -0.294907             0.387373   
google_organico              -0.086580             0.221361   
Sconti                        0.551863             0.113703   
year                         -0.362986             0.345391   
month                         0.180766            -0.150315   
day_of_week                  -0.507910            -0.278167   
day                           0.073717             0.053767   

                              google_performance_max_spent  \
revenue                                          -0.217984   
google_search_spent                               0.253990   
google_performance_max_spent                      1.000000   
fb_retargeting_spent                              0.245441   
fb_prospecting_spent                              0.309073   
google_organico                                   0.573396   
Sconti                                            0.030912   
year                                              0.497379   
month                                             0.038579   
day_of_week                                      -0.107477   
day                                               0.088795   

                              fb_retargeting_spent  fb_prospecting_spent  \
revenue                                  -0.212061             -0.294907   
google_search_spent                       0.336720              0.387373   
google_performance_max_spent              0.245441              0.309073   
fb_retargeting_spent                      1.000000              0.870278   
fb_prospecting_spent                      0.870278              1.000000   
google_organico                           0.219663              0.148569   
Sconti                                   -0.017276             -0.067479   
year                                      0.400733              0.429200   
month                                    -0.081120             -0.045990   
day_of_week                              -0.009358             -0.006461   
day                                       0.008475             -0.013582   

                              google_organico    Sconti      year     month  \
revenue                             -0.086580  0.551863 -0.362986  0.180766   
google_search_spent                  0.221361  0.113703  0.345391 -0.150315   
google_performance_max_spent         0.573396  0.030912  0.497379  0.038579   
fb_retargeting_spent                 0.219663 -0.017276  0.400733 -0.081120   
fb_prospecting_spent                 0.148569 -0.067479  0.429200 -0.045990   
google_organico                      1.000000  0.209871  0.515584  0.136474   
Sconti                               0.209871  1.000000 -0.073628  0.236032   
year                                 0.515584 -0.073628  1.000000 -0.593365   
month                                0.136474  0.236032 -0.593365  1.000000   
day_of_week                         -0.130850 -0.172997  0.000314 -0.005850   
day                                 -0.000307  0.150575 -0.036291  0.028699   

                              day_of_week       day  
revenue                         -0.507910  0.073717  
google_search_spent             -0.278167  0.053767  
google_performance_max_spent    -0.107477  0.088795  
fb_retargeting_spent            -0.009358  0.008475  
fb_prospecting_spent            -0.006461 -0.013582  
google_organico                 -0.130850 -0.000307  
Sconti                          -0.172997  0.150575  
year                             0.000314 -0.036291  
month                           -0.005850  0.028699  
day_of_week                      1.000000  0.012353  
day                              0.012353  1.000000

In [8]:
fig1 = px.scatter(data, x= 'revenue', y = 'Sconti', title = 'Revenue vs Discounts')
fig1.show()

In [9]:
#split the dataset into train and test
from sklearn.model_selection import train_test_split
train, test = train_test_split(data, test_size = 0.3, shuffle = False)

In [10]:
#Performing OLS
inputs = list(data.columns)
inputs = inputs[2:]

X= train[inputs]
y= train['revenue']
X = sm.add_constant(X)

result = sm.OLS(y, X).fit()

# printing the summary table
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                revenue   R-squared:                       0.633
Model:                            OLS   Adj. R-squared:                  0.625
Method:                 Least Squares   F-statistic:                     78.05
Date:                Fri, 08 Sep 2023   Prob (F-statistic):           5.17e-83
Time:                        01:41:36   Log-Likelihood:                -3843.1
No. Observations:                 417   AIC:                             7706.
Df Residuals:                     407   BIC:                             7746.
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------
const           

#### The OLS regression model has an R^2 of 0.633, indicating that it accounts for 63.3% of the variability in the 'revenue'. Several predictors such as 'google_search_spent', 'google_performance_max_spent', and 'day_of_week' have statistically significant coefficients based on their p-values. The model suggests potential multicollinearity issues, given the large condition number (Cond. No. ).

#### Lets plot a graph comparing actual revenue and predicted revenue using the model

In [11]:
#Comparing actual revenue to predicted revenue
Y_test = test['revenue']
X_test = test.drop(['ordine_data', 'revenue'], axis =1)

# predictions
X_test_co = sm.add_constant(X_test)
Y_pred = result.predict(X_test_co)

#creating final dataframe
Y_pred.name = 'Y_pred'
data_final = pd.concat([test, Y_pred], axis =1)
data_final



ordine_data  revenue  google_search_spent  google_performance_max_spent  \
417  2021-08-27  2266.13               136.72                         50.93   
418  2021-08-28  1728.95               102.06                         24.43   
419  2021-08-29  2668.24                96.99                         55.79   
420  2021-08-30  4974.19               172.51                         55.46   
421  2021-08-31  4830.26               139.51                         29.45   
..          ...      ...                  ...                           ...   
591  2022-02-17  6061.37               347.75                        125.32   
592  2022-02-18  3722.15               258.10                         79.30   
593  2022-02-19  2670.47                92.04                          9.96   
594  2022-02-20  3385.87                86.37                         40.50   
595  2022-02-21   372.27                42.17                         22.04   

     fb_retargeting_spent  fb_prospecting_spent  google_organico  Sconti  \
417                  0.00                  0.00                0       1   
418                  0.00                  0.00                0       0   
419                  0.00                  0.00                0       0   
420                  4.13                 34.35                0       0   
421                  1.73                 34.65                0       1   
..                    ...                   ...              ...     ...   
591                 19.96                 30.44             5003       9   
592                 20.31                 31.06             4696      14   
593                 23.16                 33.75             3058       8   
594                 19.90                 30.59             1398       3   
595                  1.33                  2.13                0       0   

     year  month  day_of_week  day       Y_pred  
417  2021      8            4   27  4979.836716  
418  2021      8            5   28  4463.300205  
419  2021      8            6   29  2869.476618  
420  2021      8            0   30  6974.632608  
421  2021      8            1   31  6502.946193  
..    ...    ...          ...  ...          ...  
591  2022      2            3   17  2965.236823  
592  2022      2            4   18  2950.176659  
593  2022      2            5   19  2359.181026  
594  2022      2            6   20   547.901766  
595  2022      2            0   21  6760.057579  

[179 rows x 13 columns]

In [12]:
#plotting actual and predicted revenue
import plotly.graph_objects as go
fig = go.Figure(go.Scatter(x = data_final['ordine_data'], y=data_final['revenue'], name = 'revenue'))
fig.add_trace(go.Scatter(x=data_final['ordine_data'], y=data_final['Y_pred'], name = 'predicted revenue'))
fig.update_layout(yaxis_title="Revenue")
fig.show()

#### From the graph, we can see that the model captures the general trend of revenue, however its precision in predicting specific revenue values falls short. Upon evaluating the OLS regression results, it's evident that addressing the evident multicollinearity, as indicated by the substantial condition number, is crucial for model improvement. There's also potential to enhance the model's significance by re-evaluating or transforming predictors with high p-values, such as 'const', 'google_organico', 'year', and 'day'. Furthermore, to better capture the underlying patterns in the data, we might consider introducing other factors not currently in the model.  We will continue working on this model to improve its performance.